schemars_derive/src/lib.rs

```diff
@@ -141,10 +141,17 @@ fn derive_json_schema(
     })
 }
 
-fn add_trait_bounds(generics: &mut syn::Generics) {
-    for param in &mut generics.params {
-        if let syn::GenericParam::Type(ref mut type_param) = *param {
-            type_param.bounds.push(parse_quote!(schemars::JsonSchema));
+fn add_trait_bounds(cont: &mut Container) {
+    if let Some(bounds) = cont.serde_attrs.ser_bound() {
+        let where_clause = cont.generics.make_where_clause();
+        where_clause.predicates.extend(bounds.iter().cloned());
+    } else {
+        // No explicit trait bounds specified, assume the Rust convention of adding the trait to each type parameter
+        // TODO consider also adding trait bound to associated types when used as fields - I think Serde does this?
```

Comment:
I haven't looked into whether Serde *tries* to do this, but so far I've always had to override Serde's bounds too when I tried to use it with associated types.

Reply:
It does automatically handle associated types, e.g. this derive:
```rust
#[derive(Serialize)]
pub struct MyContainer<T>
where
    T: Iterator,
{
    pub associated: T::Item,
    pub generic: PhantomData<T>,
}
```

expands to this impl:
```rust
impl<T> _serde::Serialize for MyContainer<T>
where
    T: Iterator,
    T::Item: _serde::Serialize,
{
```

Interestingly, it knows not to add an unnecessary trait bound on `PhantomData<T>`. [PhantomData](https://github.com/serde-rs/serde/blob/6e94a27c76834c540e36fbe8bd881a60f1ad61e9/serde_derive/src/bound.rs#L128-L134) is a special case in this regard.

I would have thought that this could be done for generic fields as well as associated types - e.g. in the above example, add a bound on `PhantomData<T>: _serde::Serialize`, which would be satisfied for all `T`, and then the `PhantomData` special case wouldn't be necessary.

Reply:
Huh, fascinating...

Reply:
Ah, looks like Serde has a special case for using `T::Item` exactly. `Option<T::Item>` (for example) still seems to require a manual bound.